script/gen_requirements_all.py

```diff
@@ -295,34 +308,44 @@ def process_requirements(errors, module_requirements, package, reqs):
 
 def generate_requirements_list(reqs):
```

Comment:
I think that instead of making an instance check, you should convert to dict BEFORE passing to the function, and adjust the signature here to:
```suggestion
def generate_requirements_list(reqs: dict[str, list[str]], separate: bool = True) -> str:
```

Reply:
The idea was that you could pass in either a bare list of requirements, or a dict that forms headers too.

Reply:
I just feel that the conversion from `list -> dict` should be moved to `requirements_output`
```python
    reqs: dict[str, list[str]] = dict.fromkeys(core_requirements(), [])
    output.append(generate_requirements_list(reqs, False))
```

Reply:
And I feel it's better for there to be one way of calling this function. You'd never want to call it with a dict that _does_ have headers and `separate=False`.